### ISSUE TITLE
Missing localization when xml:lang on child tags

### ISSUE THREAD

**Issue by Jehan** [hughsie/appstream-glib#381]
Tested with appstream-glib 0.7.17

# Issue

If the `xml:lang` properties are not on the top tag you are looking at (e.g. description), `as_app_get_description()` will not return the proper localized description. Typically is you have `xml:lang` on `<p>` tags under `<description>`. This has been verified with the `<description>` tag but I am guessing the same issue would apply on other tags which could typically have subtags.

# Reproduction steps

1. Say you have the following description:
```
  <description>
    <p>This extension provides a set of basic examples to demonstrate how to create your own plug-ins. Each plug-in does the same thing, except it is developed in a different programming language. They all create a GTK+ dialog with a text view displaying their own code (hence also demonstrating how to package data) and a button which calls a GEGL operation on the active layer.</p>
    <p xml:lang="es">Esta extensión proporciona un conjunto de ejemplos básicos para demostrar cómo crear sus propios complementos. Cada complemento hace lo mismo, solo que está desarrollado en un lenguaje de programación diferente. Todos ellos crean un diálogo GTK+ con una vista de texto que muestra su propio código (por tanto, también demuestra cómo empaquetar datos) y un botón que llama a una operación GEGL en la capa activa.</p>
    <p xml:lang="uk">До цього розширення включено набір базових прикладів. Воно є демонстрацією створення ваших власних додатків. Усі додатки виконують ті самі дії, але їх написано різними мовами програмування. Усі додатки створюють діалогове вікно GTK+ із панеллю для показу власного коду (отже, демонструють спосіб пакування даних) та кнопкою, натискання якої викликає дію GEGL над активним шаром.</p>
  </description>
```
2. Call `as_app_get_description (app, "es");`

Result: you get the whole contents of description with all 3 languages in it.

Expected result: you want to get only the Spanish translation.

# Note

If you had:

```
  <description>
    <p>This extension provides a set of basic examples to demonstrate how to create your own plug-ins. Each plug-in does the same thing, except it is developed in a different programming language. They all create a GTK+ dialog with a text view displaying their own code (hence also demonstrating how to package data) and a button which calls a GEGL operation on the active layer.</p>
  </description>
  <description xml:lang="es"> 
    <p>Esta extensión proporciona un conjunto de ejemplos básicos para demostrar cómo crear sus propios complementos. Cada complemento hace lo mismo, solo que está desarrollado en un lenguaje de programación diferente. Todos ellos crean un diálogo GTK+ con una vista de texto que muestra su propio código (por tanto, también demuestra cómo empaquetar datos) y un botón que llama a una operación GEGL en la capa activa.</p>
  </description>
  <description xml:lang="uk">
    <p>До цього розширення включено набір базових прикладів. Воно є демонстрацією створення ваших власних додатків. Усі додатки виконують ті самі дії, але їх написано різними мовами програмування. Усі додатки створюють діалогове вікно GTK+ із панеллю для показу власного коду (отже, демонструють спосіб пакування даних) та кнопкою, натискання якої викликає дію GEGL над активним шаром.</p>
  </description>
```

Then the problem would not appear. Yet this is not a solution. Why we have localization at `<p>` level is because this allows translators to work on smaller bits of strings. Also this abstracts the text layout (we don't want translators to have to care of paragraph, lists and such).
So we set the non-localized source as `<_p>` and the finale appdata is generated with merging po file translations.
Other tags can be similarly marked for translation, such as list items (`<li>`) and so on. appstream-glib should be able to just extract the relevant languages even with intertwined localized subtags.

**Comment by hughsie** [comment]
There are two forms of translations, the merged "denorm" ones with xml:lang on the <description> tags and the "source" ones with the xml:lang on the <p> tags, for example. The latter is expected in metainfo.xml files, the former expected in AppStream metadata.

**Comment by Jehan** [comment]
I am guessing there are some missing tags in your comment so I am a bit lost. Use backticks around them otherwise github will eat them out. 🙂

> The latter is expected in metainfo.xml files, the former expected in AppStream metadata.

I'm lost here too. Are you saying that metainfo.xml != AppStream metadata ?

**Comment by hughsie** [comment]
> Use backticks around them

Eeek, sorry.

> Are you saying that metainfo.xml != AppStream metadata

Sure! The former is the source that's designed to be easy to translate, the latter is the destination which is supposed to be super quick to parse and process.

**Comment by Jehan** [comment]
> Sure! The former is the source that's designed to be easy to translate, the latter is the destination which is supposed to be super quick to parse and process.

So are you saying that we do right when we localize with `<_p>` in the source, but the finale xml should have moved the xml:lang to the parent level? If so, do you have any advice on which tool to use. We have been using `intltool-merge` for years now to localize the metainfo.xml file and it certainly doesn't do this (or maybe do we miss an option?). When we have `<_p>`, it keeps the xml:lang on the finale `<p>` tags.

**Comment by hughsie** [comment]
> When we have <_p>, it keeps the xml:lang on the finale <p> tags.

Right, sorry, maybe I milead you with source and destination. My my point of view the source is the <p> post-processed metainfo.xml files used by appstream-builder, and the destination is the combined distro metadata.xml.gz

**Comment by ximion** [comment]
"AppStream Metadata" is really two specifications, one derived from the other. That's why the spec calls one "upstream metadata" or "metainfo data", and the other is called "collection metadata". The former is the stuff application authors write and is stored in metainfo files, the latter is what distributors and software centers ship and appears when a lot of component metadata is combined in one file (which can be XML or YAML, even, while metainfo files are *always* XML). You can see this in section 2 and 3: https://www.freedesktop.org/software/appstream/docs/

**Comment by Jehan** [comment]
Ah ok now I understand what you mean.

Ok just to give a bit of context: in GIMP, we will have a new extension format, which is basically just an archive containing the actual contents (can be a plug-in, brushes, patterns, etc.) and the metadata file. For the metadata, instead of creating our own stuff, I just reuse AppStream (with a few custom tags for GIMP-specific stuff, but it's 99% normal AppStream otherwise). Then extension creators will create their own metainfo.xml for their extension (which won't be installed anywhere on system side, it stays with the extension, we use the format, but these are not system application metadata, only for GIMP usage).

So I am using appstream-glib to parse these files, and now that I have one with a few localization in our repo, I realize the description is wrong (as described above). Right now I am only parsing individual metainfo files (similar to what you call "upstream metadata"), typically to get info from installed extensions.
Though I will at some point parse from collection metadata as well most likely (it's not implemented yet, but it will be possible to search and install extensions from GIMP extension repositories, in which case, I guess I may as well use the collection format as well).

So what do you suggest, regarding this issue of broken description localization?

Are you saying maybe that `as_app_get_description()` should only be used on collection metadata, not upstream metadata?

**Comment by hughsie** [comment]
> instead of creating our own stuff, I just reuse AppStream

seems sane.

>  I am using appstream-glib to parse these files

I think there is some heuristic to work out when to denorm the data. Maybe set the source kind from memory? Then as_app_get_description() should do the right thing in both cases.

**Comment by ximion** [comment]
> So I am using appstream-glib to parse these files, and now that I have one with a few localization in our repo, I realize the description is wrong (as described above). Right now I am only parsing individual metainfo files (similar to what you call "upstream metadata"), typically to get info from installed extensions.
> Though I will at some point parse from collection metadata as well most likely (it's not implemented yet, but it will be possible to search and install extensions from GIMP extension repositories, in which case, I guess I may as well use the collection format as well).

The two formats can be distinguished easily: A metainfo file has the file extension `.metainfo.xml` or `.appdata.xml` and the XML root node must be named `<component>`. A collection file can have any name, be in YAML or XML format and in the case of XML must have a root node named `<components>` with individual `<component>` nodes as children.

For your particular case, I think most of the time plain metainfo files will be good enough :-)

**Comment by Jehan** [comment]
>  Maybe set the source kind from memory?

Yeah actually it doesn't even need to be set manually. The library definitely knows the source. Like for instance, an `AsApp` which was created from an upstream metadata would be created like this `app = as_app_new (); as_app_parse_file (…);` but an `AsApp` which is part of a collection would come from a call to one of the `as_store_get_app_by_id*()` functions.
Basically when we allocate the `AsApp`, we know what its source is.

**Comment by imciner2** [comment]
I would like to revive this conversation, because this also affects running validation on `.metainfo.xml` files from upstream sources using `appstream-util validate-strict`. The metainfo spec specifies that:
> In metainfo files, this tag should be translated by-paragraph, meaning that in a translated file, each translated `<p/>` child has a language property.

Therefore, the internationalization tools that translate a metainfo file create this:
```
<description>
 <p> This is a description </p>
 <p xml:lang=fr> This is the French translation </p>
</description>
```

When I run `appstream-util validate-strict` on a metainfo file with this structure, it always gives an error saying the description field is untranslated even though there are `<p>` tags with other languages.

**Comment by Vogtinator** [comment]
I can confirm this issue.

All `/usr/share/metainfo/*.metainfo.xml` components show up with a broken description in software centers due to this, because `appstream-builder` does not convert translations in `.metainfo.xml` files to "denorm" translation in collection files (`.xml.gz`).

(Downstream report: https://bugzilla.opensuse.org/show_bug.cgi?id=1192243)

**Comment by Vogtinator** [comment]
I opened a PR: https://github.com/hughsie/appstream-glib/pull/427

I'm not very familiar with the appstream spec and a bit confused by the upstream vs. collection files already, but I can't find anything about description translation being different between `.metainfo.xml` and `.appdata.xml`.

https://www.freedesktop.org/software/appstream/docs/chap-Metadata.html#tag-description says:

> In metainfo files, this tag should be translated by-paragraph, meaning that in a translated file, each translated `<p/>` child has a language property.

And further up:

>  Component metadata of type desktop-application as described in [Section 2.2, “Desktop Applications”](https://www.freedesktop.org/software/appstream/docs/sect-Metadata-Application.html) can be installed with an .appdata.xml extension as well for historical reasons. AppStream implementations will read the XML files as long as they end up in the right location on the filesystem.

**Comment by imciner2** [comment]
My understanding is that `.metainfo.xml` and `.appdata.xml` files follow the same specification, and all that happened was the files and directories were renamed.
